### PR TITLE
EOS-8947: Decoder for ADDB logs

### PR DESCRIPTION
**Pull request by vimalraghwani20**
**Problem Statement**
EOS-8947: Decoder for ADDB logs

**Problem Description**
Implement the decoder for the ADDB logs

**Solution Overview**

- Create a plugin for decoder plugin.c 
- Implement the m0_addb2_load_interps() method which is exposed by motr dump.
- Decode the following kind
   - State 
   - Attribute
  - Map

**Unit Test Cases**
`
**sudo m0addb2dump -p /usr/lib64/libcortx-utils.so -- /addb_27527/o/100000000000000\:2 | grep fsuser** 
* 2020-09-29-17:27:47.484505618 fsuser_state     fsal_read,  state,      5035,       init
* 2020-09-29-17:27:47.484506598 fsuser_attribute fsal_read,  attribute,  5035,       read_offset, ?               0?
* 2020-09-29-17:27:47.484507284 fsuser_attribute fsal_read,  attribute,  5035,       read_vector_count, ?               1?
* 2020-09-29-17:27:47.484508075 fsuser_attribute fsal_read,  attribute,  5035,       read_length, ?            1000?
* 2020-09-29-17:27:47.484524037 fsuser_state     cfs_read,   state,      5036,       init
* 2020-09-29-17:27:47.484524850 fsuser_map       cfs_read,   map,        6,          5036,       ?            5035?
* 2020-09-29-17:27:47.484525701 fsuser_attribute cfs_read,   attribute,  5036,       read_cfs_count, ?            1000?
* 2020-09-29-17:27:47.484526538 fsuser_attribute cfs_read,   attribute,  5036,       read_cfs_offset, ?               0?
* 2020-09-29-17:27:47.484528239 fsuser_state     kvs_alloc,  state,      5037,       init
* 2020-09-29-17:27:47.484528994 fsuser_map       kvs_alloc,  map,        10,         5037,       ?            5035?
* 2020-09-29-17:27:47.484529869 fsuser_attribute kvs_alloc,  attribute,  5037,       kvs_alloc_size, ?              12?
* 2020-09-29-17:27:47.484531526 fsuser_attribute kvs_alloc,  attribute,  5037,       kvs_alloc_result, ?               0?
* 2020-09-29-17:27:47.484532228 fsuser_state     kvs_alloc,  state,      5037,       finish
* 2020-09-29-17:27:47.484533046 fsuser_state     kvs_get,    state,      5038,       init
* 2020-09-29-17:27:47.484533789 fsuser_map       kvs_get,    map,        10,         5038,       ?            5035?
* 2020-09-29-17:27:47.484534649 fsuser_attribute kvs_get,    attribute,  5038,       kvs_get_key_length, ?              12?
* 2020-09-29-17:27:47.484535507 fsuser_attribute kvs_get,    attribute,  5038,       kvs_get_value_length, ?               0?
* 2020-09-29-17:27:47.487947750 fsuser_attribute kvs_get,    attribute,  5038,       kvs_get_result, ?               0?
* 2020-09-29-17:27:47.487948775 fsuser_state     kvs_get,    state,      5038,       finish
* 2020-09-29-17:27:47.487951573 fsuser_state     cfs_get_attribute, state, 5039,     init
* 2020-09-29-17:27:47.487952315 fsuser_map       cfs_get_attribute, map, 6,          5039,       ?            5035?
* 2020-09-29-17:27:47.487953700 fsuser_state     kvs_alloc,  state,      503a,       init
* 2020-09-29-17:27:47.487954440 fsuser_map       kvs_alloc,  map,        10,         503a,       ?            5035?
* 2020-09-29-17:27:47.487955167 fsuser_attribute kvs_alloc,  attribute,  503a,       kvs_alloc_size, ?              12?
* 2020-09-29-17:27:47.487956506 fsuser_attribute kvs_alloc,  attribute,  503a,       kvs_alloc_result, ?               0?
* 2020-09-29-17:27:47.487957501 fsuser_state     kvs_alloc,  state,      503a,       finish
* 2020-09-29-17:27:47.487958297 fsuser_state     kvs_get,    state,      503b,       init
* 2020-09-29-17:27:47.487959048 fsuser_map       kvs_get,    map,        10,         503b,       ?            5035?

`

### CODE REVIEW

File: c-utils/src/addb/plugin.c [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                                const uint64_t *v, char *buf) {
+  if (v[0] == PFT_FSAL_WRITE)
+    sprintf(buf, "%s", "fsal_write");
+  if (v[0] == PFT_FSAL_READ)
+    sprintf(buf, "%s", "fsal_read");
+  if (v[0] == PFT_CFS_WRITE)
+    sprintf(buf, "%s", "cfs_write");
+  if (v[0] == PFT_CFS_READ)
+    sprintf(buf, "%s", "cfs_read");
+  if (v[0] == 	PFT_KVS_ALLOC)
```
Comment (AtitaShirwaikar) [inline c-utils/src/addb/plugin.c:134]: extra space

**Review by AtitaShirwaikar — APPROVED** [review]

File: c-utils/src/addb/plugin.c [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                                const uint64_t *v, char *buf) {
+  if (v[0] == PES_GEN_INIT)
+    sprintf(buf, "%s", "init");
+  if (v[0] == PES_GEN_FINI)
+    sprintf(buf, "%s", "finish");
+}
+
+static void perfc_entity_attrs(struct m0_addb2__context *ctx,
+                               const uint64_t *v, char *buf) {
+  if (v[0] == PEA_R_OFFSET)
```
Comment (pratyush-seagate) [inline c-utils/src/addb/plugin.c:64]: Will it be convenient to use switch case?

File: c-utils/src/addb/plugin.c [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    sprintf(buf, "%s", "kvs_get_key_length");
+  if (v[0] == PEA_KVS_GET_VLEN)
+    sprintf(buf, "%s", "kvs_get_value_length");
+  if (v[0] == PEA_KVS_GET_RES_RC)
+    sprintf(buf, "%s", "kvs_get_result");
+}
+
+static void perfc_function_tags(struct m0_addb2__context *ctx,
+                                const uint64_t *v, char *buf) {
+  if (v[0] == PFT_FSAL_WRITE)
```
Comment (pratyush-seagate) [inline c-utils/src/addb/plugin.c:126]: Same comment related to the use of switch-case

File: c-utils/src/addb/plugin.c [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+static void bol(struct m0_addb2__context *ctx, const uint64_t *v, char *buf) {
+  sprintf(buf, "%s", v[0] ? "true" : "false");
+}
+
+/* end of clip from dump.c */
+
+static void perfc_entity_states(struct m0_addb2__context *ctx,
+                                const uint64_t *v, char *buf) {
+  if (v[0] == PES_GEN_INIT)
+    sprintf(buf, "%s", "init");
```
Comment (pratyush-seagate) [inline c-utils/src/addb/plugin.c:57]: Pls use snprintf instead

**Review by ivan-alekhin — COMMENTED** [review]
If you have a mapping that maps a set of 0..N integers to a set of strings without gaps then please use a table instead "if" or "switch":

````c
enum foo_types {
  FOO_A,
  FOO_B,
  FOO_NR
};

static const struct foo_map_item {
  const char* str;
} g_foo_map[] = {
  [FOO_A] = { .str = "foo_a" },
  [FOO_B] = { .str = "foo_b" },
};
_Static_Assert(ARRAY_SIZE(g_foo_map) == FOO_NR, "foo_types-to-str should have no gaps");

static void decode(char *buf, uint64_t v)
{
  M0_PRE(v < FOO_NR);
  strcpy(buf, g_foo_map[v].str);
}
````

It reduces the amount of potential conditional jumps and makes code easier to support (for example, you can use the XLIST technique to reduce the amount of boilerplate code).

File: c-utils/src/addb/plugin.c (hunk 1 above)
Comment (vimalraghwani20) [reply c-utils/src/addb/plugin.c]: fixed.

File: c-utils/src/addb/plugin.c (hunk 2 above)
Comment (vimalraghwani20) [reply c-utils/src/addb/plugin.c]: Used the Ivan approach to avoid if and switch

File: c-utils/src/addb/plugin.c (hunk 3 above)
Comment (vimalraghwani20) [reply c-utils/src/addb/plugin.c]: Used the Ivan approach to avoid if and switch

File: c-utils/src/addb/plugin.c (hunk 4 above)
Comment (vimalraghwani20) [reply c-utils/src/addb/plugin.c]: No longer exist. used strcpy() instead.

File: c-utils/src/addb/plugin.c [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+#include "perf/perf-counters.h"
+#include <lib/assert.h>
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(a) (sizeof(a)/sizeof(a[0]))
+#endif
+
+/* Borrowed from addb2/dump.c, hope Motr will publish it as API in future */
+
+static void dec(struct m0_addb2__context *ctx, const uint64_t *v, char *buf) {
```
Comment (ivan-alekhin) [inline c-utils/src/addb/plugin.c:34]: [style] Let's keep the code style consistent and put the open curly bracket of a function definition at the next line:

```c
void foo(void)
{
}
```

Example: https://github.com/Seagate/cortx-utils/blob/main/c-utils/src/cortx/m0common.c

Let's update this and the other functions from this source file.

**Review by ivan-alekhin — COMMENTED** [review]
Overall the patch looks good to me, just one last minor comment about coding style

**Review by AtitaShirwaikar — APPROVED** [review]
Looks good to me

File: c-utils/src/addb/plugin.c (hunk 5 above)
Comment (vimalraghwani20) [reply c-utils/src/addb/plugin.c]: fixed.

**Review by ivan-alekhin — APPROVED** [review]
The patch looks good to me. Thanks for addressing the CR comments!

**Review by pratyush-seagate — APPROVED** [review]
LGTM, +1

**Review by sachinpunadikar — APPROVED** [review]
+1

**Pull request merged**